teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md

```diff
@@ -0,0 +1,537 @@
+their own choices?
+-->
+
+### User Experience
+
+The process to provide the above results could be challenging for Pipeline and Task authors. An [`oci-image-artifact-registry`](https://github.com/ywluogg/oci-image-structured-results) Task and TaskRun is provided with steps to build and push and OCI image, and then produce the above provenance infos.
+
+### Performance
+
+<!--
+(optional)
+
+Consider which use cases are impacted by this change and what are their
+performance requirements.
+- What impact does this change have on the start-up time and execution time
+of TaskRuns and PipelineRuns?
+- What impact does it have on the resource footprint of Tekton controllers
+as well as TaskRuns and PipelineRuns?
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Tekton ecosystem. Consider including folks that also work outside the WGs
+or subproject.
+- How will security be reviewed and by whom?
+- How will UX be reviewed and by whom?
+-->
+
+### Drawbacks
+
+<!--
+Why should this TEP _not_ be implemented?
+-->
+
+## Alternatives
+
+### Inputs / Outputs Distinguishment
+
+We can also separately collect inputs provenance from params, and outputs provenance from TaskRun results. The signable artifacts being found in params will be the inputs, and those found in TaskRun results will be the outputs. However, params are not reliable as users don't really specify those human less readable provenance metadata formats mentioned in in-toto attestations in neither PipelineRun nor TaskRun params.
+
+### Using Run Status to generate Provenance Metadata
+
+Results is not an ideal place to have the provenance metadata populated for the artifacts, since Results are provided by Pipeline authors. When people are looking for provenance, the metadata being provided should ideally be trustful and unchangable after generation. How can we garantee the provenance metadata can be trusted? Assuming the Tekton Pipeline being installed and where the Runs are operated are trustful, one missing piece to comply to SLSA level L3 is that the Run yaml being submitted is a trustful config, which can be fulfilled by [TEP 091: Trusted Resource](https://github.com/tektoncd/community/pull/739). The other missing piece is that we want to make sure the generated provenance data shouldn't be changable by any untrusted threads. TaskRun and PipelineRun `status` field satisfies these needs, as only Pipeline controller is able to generate info within this field. To extend the previous trustful setup, we can allow Trusted Tasks to modify the field as well: when Trusted Tasks are being used, Pipeline Controller can verify the Trusted Tasks being used in a Run, and let Trusted Tasks to generate provenance metadata a new field under `status` fields.
```

Comment:
I think the thing I like the most about not tying provenance to to results is that authors don't need to pre-declare provenance result data - a Task can simply output it without needing it to be explicitly declared in a param. This makes it easier to work with when there's a variable amount of items that will be generated, though at the cost of making it hard to be consumed by other steps (though it's unclear to me whether there's a strong requirement for this at the moment).

Reply:
I think if users want another step to verify if the produced result is having the correct digest or verify if the uploaded digest matches the computed digest, this becomes critical that the provenance can be consumed in another step

Reply:
I still find it useful to have provenance metadata retrieval being supported in results, since in status field that would require Trusted Resources, but I see circumstances that users may not be able to come all the way to support Trusted Resources. Having this supported in results can give users more flexibility to whether to choose Trusted Resources or not

Reply:
My thinking is that Tasks could still produce output data into results - it just wouldn't be a requirement and authors can format that data however they want. The benefit is splitting the provenance data that has a predefined API that is meant to be used by other tools from Task results that are meant to be used by other human authors.

We can still expose the raw data into Tasks later on if we wanted to, but we also don't need to commit to that right now.

Reply:
I see! That's a good point. But how would we allow users to use Chains, if the string format type hinting is something we are looking to deprecate? I think users may still look for a way to use Chains in a more structured way without using Trusted Resources

Reply:
> But how would we allow users to use Chains, if the string format type hinting is something we are looking to deprecate?

How confident are we in the status based provenance as the long term solution? It might make more sense to just move forward with that.
We could support a limited subset short term if needed (i.e. single, predefined artifacts) with the understanding that it's not a general purpose solution and can probably only be used in limited cases.
Since it sounds like we need Pipeline changes anyway to have a more general solution using params (i.e. object arrays), the difference between that and adding a new status field on the surface doesn't feel like that big of a delta.

> I think the field we plan to add in status requires the Task to be trusted task

Does it? It doesn't seem that much different than the parameter based approach - it just gives us a more controlled/structured location to put provenance data in Run objects.

Reply:
Discussed offline with @wlynch and updated a note at the top of Proposal section!